File: DailyTemperature.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None

public class DailyTemperature {
    class Solution {
        public int[] dailyTemperatures(int[] temperatures) {
            int n = temperatures.length;
            Stack<Integer> stack = new Stack<>();
            stack.push(0);
            int[] result = new int[n];

            for(int i = 1; i < n; i++){
                while(!stack.isEmpty() && temperatures[i] > temperatures[stack.peek()]){
                    int popped = stack.pop();
                    result[popped] = i - popped;
                }
                stack.push(i);
            }
            return result;
        }
    }
}

File: dailyTemperature.cpp
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes

// Three line explanation of solution in plain english
/* Create a stack to store index of i/p vector and result vector to store answer.
 * Iterate over i/p vector, if top element of stack is less than the current element then
 * update current index in result vector by subtracting stack top element from current index.
 * Repeat the above step until stack is empty or greater element in stack is encountered.
 * Return result vector at the end.  
 */

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        if (temperatures.size() <= 0)
        {
            return {};
        }
        
        stack<int> stk;
        vector<int> result(temperatures.size(), 0);
        
        for (int i = 0; i < temperatures.size(); i++)
        {
            while (!stk.empty() && temperatures[stk.top()] < temperatures[i])
            {
                result[stk.top()] = i - stk.top();
                stk.pop();
            }
            
            stk.push(i);
        }
        
        return result;
    }
};